### PR TITLE
HPCC-18419 Service to retrieve serialized typeInfo information

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by richardkchapman** [comment]
Commit titles need fixing - and please squash

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 1]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
     return true;
 }
 
+bool CWsDfuEx::onDFURecordTypeInfo(IEspContext &context, IEspDFURecordTypeInfoRequest &req, IEspDFURecordTypeInfoResponse &resp)
+{
+    try
+    {
+        if (!context.validateFeatureAccess(FEATURE_URL, SecAccess_Read, false))
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "Failed to access DFUDefFile. Permission denied.");
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Wrong method name in exception text.

File: esp/scm/ws_dfu.ecm [hunk 2]
```diff
@@ -319,6 +323,17 @@ DFUDefFileResponse
     [http_content("application/octet-stream")] binary defFile;
 };
 
+ESPrequest DFURecordTypeInfoRequest
+{
+    string Name;
+};
```
Comment (afishbeck) [inline esp/scm/ws_dfu.ecm]: Should we support:

  [min_ver("1.37")] bool IncludeJsonTypeInfo(true);
  [min_ver("1.37")] bool IncludeBinTypeInfo(true);

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 3]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+
+        const char* fileName = req.getName();
+        if (!fileName || !*fileName)
+            throw MakeStringException(ECLWATCH_MISSING_PARAMS, "File name required");
+        PROGLOG("DFUDefFile: %s", fileName);
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Wrong method name.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 4]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+
+        StringBuffer username;
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: could just use a const char * and context.queryUserName() (or whatever it's called).

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 5]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+        context.getUserID(username);
+
+        StringBuffer rawStr,returnStr;
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Unused.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 6]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+        {
+            userdesc.setown(createUserDescriptor());
+            userdesc->set(username.str(), context.queryPassword(), context.querySessionToken(), context.querySignature());
+        }
+
+        Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(fileName, userdesc);
+        if(!df)
+            throw MakeStringException(ECLWATCH_FILE_NOT_EXIST,"Cannot find file %s.",fileName);
+        if(!df->queryAttributes().hasProp("ECL"))
+            throw MakeStringException(ECLWATCH_MISSING_PARAMS,"No record definition for file %s.",fileName);
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: "ECL record definition" or just "ECL definition".

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 7]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+        text.append(df->queryAttributes().queryProp("ECL"));
+
+        MultiErrorReceiver errs;
+        OwnedHqlExpr record = parseQuery(text.str(), &errs);
+        if (errs.errCount())
+        {
+            StringBuffer errtext;
+            IError *first = errs.firstError();
+            first->toString(errtext);
+            throw MakeStringException(ECLWATCH_CANNOT_PARSE_ECL_QUERY, "Failed in parsing ECL query: %s @ %d:%d.", errtext.str(), first->getColumn(), first->getLine());
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Change "query" to "record definition".

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 8]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+        }
+
+        if(!record)
+            throw MakeStringException(ECLWATCH_CANNOT_PARSE_ECL_QUERY, "Failed in parsing ECL query.");
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: same.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 9]
```diff
@@ -2186,6 +2249,38 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor* udesc, co
...
+        if (ecltext && *ecltext)
+        {
+            MultiErrorReceiver errs;
+            OwnedHqlExpr record = parseQuery(ecltext, &errs);
+            if (errs.errCount())
+            {
+                StringBuffer errtext;
+                IError *first = errs.firstError();
+                first->toString(errtext);
+                throw MakeStringException(ECLWATCH_CANNOT_PARSE_ECL_QUERY, "Failed in parsing ECL query: %s @ %d:%d.", errtext.str(), first->getColumn(), first->getLine());
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: same.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 10]
```diff
@@ -2186,6 +2249,38 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor* udesc, co
...
+            }
+
+            if(!record)
+                throw MakeStringException(ECLWATCH_CANNOT_PARSE_ECL_QUERY, "Failed in parsing ECL query.");
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: same.

File: esp/services/ws_dfu/ws_dfuService.hpp [hunk 11]
```diff
@@ -131,7 +131,8 @@ class CWsDfuEx : public CWsDfu
     bool onDFUBrowseData(IEspContext &context, IEspDFUBrowseDataRequest &req, IEspDFUBrowseDataResponse &resp);
     bool onDFUSearchData(IEspContext &context, IEspDFUSearchDataRequest &req, IEspDFUSearchDataResponse &resp);
     bool onDFUGetFileMetaData(IEspContext &context, IEspDFUGetFileMetaDataRequest &req, IEspDFUGetFileMetaDataResponse &resp);
-
+	bool onDFURecordTypeInfo(IEspContext &context, IEspDFURecordTypeInfoRequest &req, IEspDFURecordTypeInfoResponse &resp);
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.hpp]: indentation.

**Review by afishbeck — CHANGES_REQUESTED** [review]

**Comment by afishbeck** [comment]
@shamser back to you.

**Review by afishbeck — APPROVED** [review]
@shamser looks good

**Comment by shamser** [comment]
@ghalliday it's ready to merge.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 12]
```diff
@@ -1531,6 +1533,67 @@ bool CWsDfuEx::onDFUDefFile(IEspContext &context,IEspDFUDefFileRequest &req, IEs
...
+        Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(fileName, userdesc);
+        if(!df)
+            throw MakeStringException(ECLWATCH_FILE_NOT_EXIST,"Cannot find file %s.",fileName);
+        if(!df->queryAttributes().hasProp("ECL"))
+            throw MakeStringException(ECLWATCH_MISSING_PARAMS,"No ECL record definition for file %s.",fileName);
+
+        StringBuffer text;
+        text.append(df->queryAttributes().queryProp("ECL"));
+
+        MultiErrorReceiver errs;
```
Comment (ghalliday) [inline esp/services/ws_dfu/ws_dfuService.cpp]: This code should really be shared with the code below - a bit too much cut and paste.

**Comment by shamser** [comment]
@ghalliday Code modified as per code review comment.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 13]
```diff
@@ -2251,34 +2236,18 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor* udesc, co
     }
     if (includeJsonTypeInfo||includeBinTypeInfo)
     {
-        const char *ecltext = df->queryAttributes().queryProp("ECL");
-        if (ecltext && *ecltext)
+        OwnedHqlExpr record = getEclRecordDefinition(df->queryAttributes().queryProp("ECL"));
```
Comment (ghalliday) [inline esp/services/ws_dfu/ws_dfuService.cpp]: This will now crash if queryProp("ECL") is null

**Comment by ghalliday** [comment]
@shamser see comment above

**Pull request merged**